modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,298 @@
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Reducer of distirbuted query that sort result through all nodes. Note that it's assumed that every node
+ * returns pre-sorted collection of data.
+ */
+public class MergeSortDistributedCacheQueryReducer<R> extends UnsortedDistributedCacheQueryReducer<R> {
+    /** Map of Node ID to stream. Used for inserting new pages. */
+    private final Map<UUID, NodePageStream> streamsMap;
+
+    /** Array of streams. Used for iteration over result pages. */
+    private final NodePageStream[] streams;
+
+    /** If {@code true} then there wasn't call {@link #hasNext()} on this reducer. */
+    private boolean first = true;
+
+    /**
+     * Offset of current stream to get next value. This offset only increments if a stream is done. The array {@link
+     * #streams} is sorted to put stream with lowest value on this offset.
+     */
+    private int streamOff;
+
+    /** Compares head of streams to get lowest value at the moment. */
+    private final Comparator<NodePageStream> streamCmp;
+
+    /**
+     * @param rowCmp Comparator to sort query results from different nodes.
+     */
+    public MergeSortDistributedCacheQueryReducer(GridCacheQueryFutureAdapter fut, long reqId,
+        CacheQueryPageRequester fetcher, Collection<ClusterNode> nodes, Comparator<R> rowCmp
+    ) {
+        super(fut, reqId, fetcher, nodes);
+
+        synchronized (queueLock()) {
+            streamsMap = new ConcurrentHashMap<>(nodes.size());
+            streams = (NodePageStream[])Array.newInstance(NodePageStream.class, nodes.size());
+
+            int i = 0;
+
+            for (ClusterNode node : nodes) {
+                streams[i] = new NodePageStream(node.id());
+                streamsMap.put(node.id(), streams[i++]);
+            }
+        }
+
+        streamCmp = (o1, o2) -> {
+            if (o1 == o2)
+                return 0;
+
+            if (o1 == null)
+                return -1;
+
+            if (o2 == null)
+                return 1;
+
+            return rowCmp.compare(o1.head, o2.head);
+        };
+    }
+
+    /** {@inheritDoc} */
+    @Override public R next() throws IgniteCheckedException {
+        for (int i = streamOff; i < streams.length; i++) {
+            NodePageStream s = streams[i];
+
+            // Skip if stream has head: on previous next() head extracts from stream[streamOff], others streams do
+            // have head or it is the first run.
+            if (s.head() != null)
+                break;
```

Comment:
AFAIK, at the very first call, first pages will be requested from all data nodes.
The next data will be got from the head.
When the first page data will be consumed, then the next page will be requested synchronously.

Thus, prefetch only works for the first nodes pages only.

Reply:
Next pages are requested asynchronously in moment after preparing a head for a stream and finding that queue is empty. We compare heads and bubble them on every user's `next()`. So for multiple streams (1 stream per node), for even-distributed between nodes data we will get head from new stream for every next() call. Then we can expect that page loading is non-blocking in most cases.

Reply:
Check this code:
https://github.com/apache/ignite/pull/9081/files#diff-068da2d055341b4686f2dc28d687b5c58678a255cde712e4062588bf0fc4f46fR135

Reply:
After you got the last item from the first page, you will be blocked on `next()` call until the second page will be received from the node, otherwise you will not be able to resort streams correctly.
You need to have at least one page from every node to be ready at any time for non-blocking reducing.

I wouldn't bet on equal data distribution. Assume, one node will have the largest part of the query result,
then you will end up in synchronous ping-pong with a single node.

Effective prefetching is to request next page right after you start processing the previous one.
If you bother about fetching few additional pages, then you can configure twice smaller page, aren't you?

Reply:
> Effective prefetching is to request next page right after you start processing the previous one

It is how it's done now. By "queue is empty" I mean the queue variable in PageStream, that stores page's collections. When the queue is empty it means that we start processing last page (and data moved from queue to iterator [1]) and we request a new page in parallel. 

By the link above I trigger this load right after moving page to iterator, check the condition: queue.size() < 1. So queue is empty but we still have data in iterator.

We can change this condition to queue.size() < 2, so we will have at least 2 pages in any moment (1 in iterator, 1 in a queue). So it can be useful if user processing page in iterator faster then remote node responses.

Am I right that you propose to do this (increment 1 to 2)? Should we do it configurable?

[1] https://github.com/apache/ignite/pull/9081/files#diff-068da2d055341b4686f2dc28d687b5c58678a255cde712e4062588bf0fc4f46fR75

Reply:
Reworked this, now there is a single logic for loading pages for PageStream and NodePageStream. After moving last item from queue to iterator, we trigger load page.